src/main/java/org/hydev/mcpm/client/Downloader.java

```diff
@@ -40,8 +83,45 @@ public void downloadFile(String url, File to)
      */
     public void downloadFiles(Map<String, File> urls, boolean progress, int threads)
     {
-        // TODO: Implement this
-        throw new UnsupportedOperationException("TODO");
+        Thread[] threadss = new Thread[threads];
+        final int UrlPerThread = urls.size() / threads;
+        final int remainingUrl = urls.size() % threads;
+
+        for (int t = 0; t < threads; t++) {
+            final int thread = t;
+            threadss[t] = new Thread() {
+                @Override
+                public void run() {
+                    runThread(urls, threads, thread, UrlPerThread, remainingUrl);
+                }
+            };
+        }
+
+        for (Thread t: threadss) {
+            t.start();
+        }
+    }
+    private void runThread(Map<String, File> urls, int threads, int thread, int UrlPerThread, int remainingUrl) {
+        // Store all urls into a list
+        List<String> files = new ArrayList<>();
+        for (Map.Entry<String, File> url:urls.entrySet()) {
+            files.add(url.getKey());
+        }
+        // Store urls per thread
+        List<String> inFiles = new ArrayList<>();
+
+        for (int i=thread*UrlPerThread; i < (thread + 1) * UrlPerThread; i++) {
+            inFiles.add(files.get(i));
+        }
+        if (thread == threads-1 && remainingUrl > 0) {
```

Comment:
Just out of curiosity, what does `thread == threads-1` do here?

Reply:
there are a total of `threads` threads, so `thread == threads - 1` tests if this is the final iteration of the loop

Reply:
What loop? This method doesn't call itself does it...

Reply:
I have a reading issue. But this method is called from a loop, which uses the same naming conventions, so I guess this means if this is the "final" thread

Reply:
It's for if the # files isn't divisible by # threads, to add the remainder manually